File: clients/go/internal/pathways/pathway.go
package pathways

import (
	"database/sql"
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
	"sync"

	"daisugi-verify/internal/pmodel"
	"daisugi-verify/internal/pyjson"
	"daisugi-verify/internal/pystr"
)

// ErrUnreadable is a row this binary cannot read the way Python does.
// The command refuses rather than guess.
var ErrUnreadable = errors.New("cannot read")

// Invalid is a row Python fails to read: pydantic or json raises, and
// the CLI exits 1 with a traceback.
type Invalid struct{ Msg string }

func (e *Invalid) Error() string { return e.Msg }

// Pathway is a validated CompiledPathway: model_dump() in field order.
type Pathway struct {
	Obj     *pyjson.Object
	embText string
}

// ID is the pathway's id.
func (p *Pathway) ID() string { return p.Obj.Value("id").(string) }

// Task is task_description.
func (p *Pathway) Task() string { return p.Obj.Value("task_description").(string) }

// pyValue is a column value as Python's sqlite3 gives it, in the pyjson
// types pmodel reads.
func pyValue(v any) (any, error) {
	switch x := v.(type) {
	case nil:
		return nil, nil
	case int64:
		return pyjson.Int{Text: strconv.FormatInt(x, 10)}, nil
	case float64:
		return pyjson.Float(x), nil
	case string:
		return x, nil
	}
	// A BLOB is bytes in Python; pydantic's str and float read bytes by
	// rules this port does not carry.
	return nil, fmt.Errorf("%w: a column holds a %T", ErrUnreadable, v)
}

func field(title string, s pmodel.Schema, v any) (any, error) {
	out, verr := pmodel.Validate(title, s, v, pmodel.Python)
	if verr != nil {
		return nil, &Invalid{"pydantic_core._pydantic_core.ValidationError: " + verr.String()}
	}
	return out, nil
}

func loads(text any) (any, error) {
	s, ok := text.(string)
	if !ok {
		// json.loads of an int or None raises TypeError.
		return nil, &Invalid{"TypeError: the JSON object must be str, bytes or bytearray"}
	}
	v, err := pyjson.Loads(s)
	if err != nil {
		return nil, &Invalid{"json.decoder.JSONDecodeError: " + err.Error()}
	}
	return v, nil
}

// FromRow is PathwayStore._row_to_pathway: the JSON columns decoded, the
// envelope and plan read with model_validate_json, and CompiledPathway
// built from them.
func FromRow(r Row) (*Pathway, error) {
	get := func(col string) (any, error) {
		v, present := r[col]
		if !present {
			return nil, fmt.Errorf("%w: the row has no %s column", ErrUnreadable, col)
		}
		return pyValue(v)
	}
	out := pyjson.NewObject()
	for _, col := range []string{"id", "task_description"} {
		v, err := get(col)
		if err != nil {
			return nil, err
		}
		if v, err = field("CompiledPathway", pmodel.Str{}, v); err != nil {
			return nil, err
		}
		out.Set(col, v)
	}
	embText, err := get("task_embedding_json")
	if err != nil {
		return nil, err
	}
	// json.loads and list[float]: checked here, the floats themselves
	// built only when a command prints them (Full).
	embStr, isStr := embText.(string)
	if !isStr {
		if _, err := loads(embText); err != nil {
			return nil, err
		}
	}
	if _, ok := scanEmbedding(embStr); !ok {
		if _, err := loads(embStr); err != nil {
			return nil, err
		}
		return nil, &Invalid{"pydantic_core._pydantic_core.ValidationError: " + "task_embedding is not a list of numbers"}
	}
	out.Set("task_embedding", nil)
	for _, col := range []string{"embedding_model", "embedding_model_version"} {
		v, err := get(col)
		if err != nil {
			return nil, err
		}
		if v, err = field("CompiledPathway", pmodel.Str{}, v); err != nil {
			return nil, err
		}
		out.Set(col, v)
	}
	for _, c := range []struct {
		col, key string
		model    *pmodel.Model
	}{{"envelope_json", "envelope", pmodel.Envelope}, {"plan_template_json", "plan_template", pmodel.ActionPlan}} {
		v, err := get(c.col)
		if err != nil {
			return nil, err
		}
		s, ok := v.(string)
		if !ok {
			return nil, &Invalid{"pydantic_core._pydantic_core.ValidationError: " + "model_validate_json needs a str"}
		}
		obj, verr := pmodel.ValidateJSON(c.model.Name, c.model, s)
		if verr != nil {
			for _, e := range verr.Errs {
				if e.Type == pmodel.UnreadableStep {
					return nil, fmt.Errorf("%w: a plan step is a string", ErrUnreadable)
				}
			}
			return nil, &Invalid{"pydantic_core._pydantic_core.ValidationError: " + verr.String()}
		}
		out.Set(c.key, obj)
	}
	traces, err := get("source_trace_ids_json")
	if err != nil {
		return nil, err
	}
	if traces, err = loads(traces); err != nil {
		return nil, err
	}
	if traces, err = field("CompiledPathway", pmodel.List{Elem: pmodel.Str{}}, traces); err != nil {
		return nil, err
	}
	out.Set("source_trace_ids", traces)
	for _, c := range []struct {
		col string
		s   pmodel.Schema
	}{{"version", pmodel.Int{}}, {"hit_count", pmodel.Int{}}, {"distilled_at", pmodel.Float{}}} {
		v, err := get(c.col)
		if err != nil {
			return nil, err
		}
		if v, err = field("CompiledPathway", c.s, v); err != nil {
			return nil, err
		}
		out.Set(c.col, v)
	}
	// The later columns: present on every migrated table.
	for _, c := range []struct {
		col string
		s   pmodel.Schema
	}{{"last_activation_at", pmodel.Float{}}, {"failure_count", pmodel.Int{}}} {
		v, err := get(c.col)
		if err != nil {
			return nil, err
		}
		if v, err = field("CompiledPathway", c.s, v); err != nil {
			return nil, err
		}
		out.Set(c.col, v)
	}
	out.Set("activation_count", pyjson.Int{Text: "0"})
	sig, err := get("structure_signature")
	if err != nil {
		return nil, err
	}
	if sig, err = field("CompiledPathway", pmodel.Nullable{Inner: pmodel.Str{}}, sig); err != nil {
		return nil, err
	}
	out.Set("structure_signature", sig)
	params, err := get("parameters_json")
	if err != nil {
		return nil, err
	}
	// `json.loads(p) if p else []`: an empty text or a NULL is no
	// parameters.
	var plist any = []any{}
	if pyjson.Truthy(params) {
		if plist, err = loads(params); err != nil {
			return nil, err
		}
	}
	if plist, err = field("CompiledPathway", pmodel.List{Elem: pmodel.PathwayParameter}, plist); err != nil {
		return nil, err
	}
	out.Set("parameters", plist)
	return &Pathway{Obj: out, embText: embStr}, nil
}

// Full sets task_embedding to its floats, for a command that prints it.
func (p *Pathway) Full() *Pathway {
	if p.Obj.Value("task_embedding") != nil {
		return p
	}
	vec, _ := parseEmbedding(p.embText)
	xs := make([]any, len(vec))
	for i, x := range vec {
		xs[i] = x
	}
	p.Obj.Set("task_embedding", xs)
	return p
}

// pydantic's JSON, written from a model_dump() value.

func pydanticValue(v any) any {
	switch x := v.(type) {
	case float64:
		return pyjson.Raw(floatJSON(x))
	case pyjson.Float:
		return pyjson.Raw(floatJSON(float64(x)))
	case []any:
		out := make([]any, len(x))
		for i, e := range x {
			out[i] = pydanticValue(e)
		}
		return out
	case *pyjson.Object:
		out := pyjson.NewObject()
		for _, k := range x.Keys() {
			out.Set(k, pydanticValue(x.Value(k)))
		}
		return out
	}
	return v
}

// floatJSON is pydantic's text for a float: its own exponent form, and
// null for NaN and the infinities.
func floatJSON(f float64) string {
	if math.IsNaN(f) || math.IsInf(f, 0) {
		return "null"
	}
	s := pyjson.FloatRepr(f)
	if mant, ok := strings.CutSuffix(s, "e-05"); ok {
		sign := ""
		if strings.HasPrefix(mant, "-") {
			sign, mant = "-", mant[1:]
		}
		return sign + "0.0000" + strings.Replace(mant, ".", "", 1)
	}
	if i := strings.IndexByte(s, 'e'); i >= 0 {
		mant, exp := s[:i], s[i+1:]
		sign := ""
		if exp[0] == '+' || exp[0] == '-' {
			sign, exp = exp[:1], exp[1:]
		}
		exp = strings.TrimLeft(exp, "0")
		if exp == "" {
			exp = "0"
		}
		s = mant + "e" + sign + exp
	}
	return s
}

// DumpJSON is model_dump_json(): compact, non-ASCII kept.
func DumpJSON(v any) string {
	var b strings.Builder
	compact(&b, pydanticValue(v))
	return b.String()
}

// DumpJSONIndent is model_dump_json(indent=n).
func DumpJSONIndent(v any, n int) string {
	return pyjson.DumpsIndent(pydanticValue(v), n, false)
}

func compact(b *strings.Builder, v any) {
	switch x := v.(type) {
	case []any:
		b.WriteByte('[')
		for i, e := range x {
			if i > 0 {
				b.WriteByte(',')
			}
			compact(b, e)
		}
		b.WriteByte(']')
	case *pyjson.Object:
		b.WriteByte('{')
		for i, k := range x.Keys() {
			if i > 0 {
				b.WriteByte(',')
			}
			b.WriteString(pyjson.Dumps(keyText(k), false))
			b.WriteByte(':')
			compact(b, x.Value(k))
		}
		b.WriteByte('}')
	default:
		b.WriteString(pyjson.Dumps(v, false))
	}
}

// ReadAll is list_all(): every row read as a pathway, in table order.
// Row ranges are read and checked on separate connections at once. keep
// names the pathways whose embedding a command will print; the others
// drop its text once it is checked. The error is the first row's in
// table order, as Python's loop meets it.
func (s *Store) ReadAll(keep func(id string) bool) ([]*Pathway, error) {
	type result struct {
		p   *Pathway
		err error
	}
	var mu sync.Mutex
	byPart := map[int][]result{}
	_, err := s.scanParts("*", func(part int, rows *sql.Rows) error {
		cols, err := rows.Columns()
		if err != nil {
			return err
		}
		var out []result
		for rows.Next() {
			vals := make([]any, len(cols))
			ptrs := make([]any, len(cols))
			for i := range vals {
				ptrs[i] = &vals[i]
			}
			if err := rows.Scan(ptrs...); err != nil {
				return err
			}
			r := make(Row, len(cols))
			for i, c := range cols {
				r[c] = vals[i]
			}
			s.asMigrated(r)
			p, err := FromRow(r)
			if err == nil && (keep == nil || !keep(p.ID())) {
				p.embText = ""
			}
			out = append(out, result{p, err})
		}
		mu.Lock()
		byPart[part] = out
		mu.Unlock()
		return nil
	})
	if err != nil {
		return nil, err
	}
	var all []*Pathway
	for k := 0; k < len(byPart); k++ {
		for _, r := range byPart[k] {
			if r.err != nil {
				return nil, r.err
			}
			all = append(all, r.p)
		}
	}
	return all, nil
}

// keyText is a dict key as pydantic's JSON writes it: a lone surrogate
// becomes its three UTF-8-style bytes, which read back as three U+FFFD.
func keyText(k string) string {
	if !pystr.HasSurrogate(k) {
		return k
	}
	var b strings.Builder
	for i := 0; i < len(k); {
		r, n := pystr.DecodeRune(k[i:])
		if pystr.IsSurrogate(r) {
			b.WriteString("\uFFFD\uFFFD\uFFFD")
		} else {
			b.WriteString(k[i : i+n])
		}
		i += n
	}
	return b.String()
}

File: clients/rust/src/pathways/pathway.rs
//! A stored row read as `PathwayStore._row_to_pathway` reads it, and a
//! pathway written back as pydantic writes it.

use rusqlite::types::ValueRef;

use super::pmodel::{self, take, Id, Mode, Schema};
use super::scan::{parse_embedding, scan_embedding};
use super::store::{sql_err, text, Col, Row, Store};
use super::PwErr;
use crate::gate::py::text::{as_surrogate, has_surrogate};
use crate::gate::pyjson::{dumps, dumps_indent_with, float_repr, loads, Object, Value};

/// A validated CompiledPathway: its `model_dump()`, fields in order.
/// `task_embedding` is Null until a command needs its floats (`full`).
#[derive(Debug, Clone)]
pub struct Pathway {
    pub obj: Object,
    emb_text: Option<String>,
}

impl Pathway {
    pub fn new(obj: Object) -> Pathway {
        Pathway { obj, emb_text: None }
    }

    pub fn id(&self) -> &str {
        self.obj.value("id").as_str().unwrap_or("")
    }

    pub fn task(&self) -> &str {
        self.obj.value("task_description").as_str().unwrap_or("")
    }

    /// Sets task_embedding to its floats, for a command that prints it.
    pub fn full(mut self) -> Pathway {
        if !self.obj.value("task_embedding").is_null() {
            return self;
        }
        let vec = self.emb_text.as_deref().and_then(parse_embedding).unwrap_or_default();
        self.obj.set("task_embedding", Value::List(vec.into_iter().map(Value::Float).collect()));
        self
    }
}

fn invalid_validation(e: pmodel::ValidationError) -> PwErr {
    if e.unreadable_step() {
        return PwErr::Unreadable("a plan step is a string".into());
    }
    PwErr::Invalid(format!("pydantic_core._pydantic_core.ValidationError: {}", e.text()))
}

fn col_value(c: &Col) -> Result<Value, PwErr> {
    Ok(match c {
        Col::Null => Value::Null,
        Col::Int(i) => Value::Int(i.to_string()),
        Col::Real(f) => Value::Float(*f),
        Col::Text(s) => Value::Str(s.clone()),
        // A BLOB is bytes in Python; pydantic's str and float read bytes
        // by rules this port does not carry.
        Col::Blob => return Err(PwErr::Unreadable("a column holds a BLOB".into())),
    })
}

/// `json.loads(v)`: a str is parsed; any other type raises TypeError.
fn json_loads(v: &Value) -> Result<Value, PwErr> {
    match v {
        Value::Str(s) => loads(s).map_err(|_| PwErr::Invalid("json.decoder.JSONDecodeError: the text is not JSON".into())),
        _ => Err(PwErr::Invalid("TypeError: the JSON object must be str, bytes or bytearray".into())),
    }
}

/// The embedding column, checked on its borrowed bytes.
#[derive(Debug, Clone)]
pub enum Emb {
    /// A TEXT value: whether it is a list of numbers, and its text when a
    /// command needs it (or when it is not).
    Text { ok: bool, text: Option<String> },
    Other(Col),
}

impl Emb {
    pub fn read(v: ValueRef<'_>, keep: bool) -> Result<Emb, PwErr> {
        match v {
            ValueRef::Text(b) => {
                let ok = scan_embedding(b).is_some();
                let t = if keep || !ok { Some(text(b)?) } else { None };
                if !ok {
                    // Undecodable text is still an error, as Python raises.
                }
                Ok(Emb::Text { ok, text: t })
            }
            other => Ok(Emb::Other(Col::from_ref(other)?)),
        }
    }
}

/// `_row_to_pathway`, in Python's order: the embedding's json.loads, the
/// envelope and the plan read with model_validate_json, the other JSON
/// columns loaded, then CompiledPathway's own validation of each field.
pub fn from_row(r: &Row, emb: &Emb) -> Result<Pathway, PwErr> {
    let get = |name: &str| -> Result<Value, PwErr> {
        match r.get(name) {
            Some(c) => col_value(c),
            None => Err(PwErr::Unreadable(format!("the row has no {name} column"))),
        }
    };
    // json.loads(task_embedding_json).
    let (emb_ok, emb_text) = match emb {
        Emb::Text { ok, text } => {
            if !ok {
                json_loads(&Value::Str(text.clone().unwrap_or_default()))?;
            }
            (*ok, text.clone())
        }
        Emb::Other(c) => {
            json_loads(&col_value(c)?)?;
            (false, None)
        }
    };
    let mut models = vec![];
    for (col, id) in [("envelope_json", Id::Envelope), ("plan_template_json", Id::ActionPlan)] {
        let v = get(col)?;
        let s = match v {
            Value::Str(s) => s,
            _ => {
                return Err(PwErr::Invalid(
                    "pydantic_core._pydantic_core.ValidationError: JSON input should be string, bytes or bytearray".into(),
                ))
            }
        };
        models.push(pmodel::validate_json(id, &s).map_err(invalid_validation)?);
    }
    let traces = json_loads(&get("source_trace_ids_json")?)?;
    let params_raw = get("parameters_json")?;
    let params = if params_raw.truthy() { json_loads(&params_raw)? } else { Value::List(vec![]) };

    // CompiledPathway(...): each field in its order.
    let field = |s: &Schema, v: Value| -> Result<Value, PwErr> {
        take("CompiledPathway", s, v, Mode::Python).map_err(invalid_validation)
    };
    let mut out = Object::new();
    out.set("id", field(&Schema::Str, get("id")?)?);
    out.set("task_description", field(&Schema::Str, get("task_description")?)?);
    if !emb_ok {
        return Err(PwErr::Invalid(
            "pydantic_core._pydantic_core.ValidationError: task_embedding is not a list of numbers".into(),
        ));
    }
    out.set("task_embedding", Value::Null);
    out.set("embedding_model", field(&Schema::Str, get("embedding_model")?)?);
    out.set("embedding_model_version", field(&Schema::Str, get("embedding_model_version")?)?);
    let mut it = models.into_iter();
    out.set("envelope", it.next().unwrap_or(Value::Null));
    out.set("plan_template", it.next().unwrap_or(Value::Null));
    out.set("source_trace_ids", field(&Schema::List(Box::new(Schema::Str)), traces)?);
    out.set("version", field(&Schema::Int, get("version")?)?);
    out.set("hit_count", field(&Schema::Int, get("hit_count")?)?);
    out.set("distilled_at", field(&Schema::Float, get("distilled_at")?)?);
    out.set("last_activation_at", field(&Schema::Float, get("last_activation_at")?)?);
    out.set("failure_count", field(&Schema::Int, get("failure_count")?)?);
    out.set("activation_count", Value::Int("0".into()));
    out.set("structure_signature", field(&Schema::Nullable(Box::new(Schema::Str)), get("structure_signature")?)?);
    out.set("parameters", field(&Schema::List(Box::new(Schema::Model(Id::PathwayParameter))), params)?);
    Ok(Pathway { obj: out, emb_text })
}

impl Store {
    /// `list_all()`: every row read as a pathway, in table order. `keep`
    /// names the pathways whose embedding a command will print. The error
    /// is the first row's in table order, as Python's loop meets it.
    pub fn read_all(&self, keep: &(dyn Fn(&str) -> bool + Sync)) -> Result<Vec<Pathway>, PwErr> {
        let cols = self.all_columns();
        let parts = self.scan_parts(&cols, &|rows| {
            let mut out: Vec<Result<Pathway, PwErr>> = vec![];
            let names: Vec<String> = rows.as_ref().map(|s| s.column_names().iter().map(|n| n.to_string()).collect()).unwrap_or_default();
            let ei = names.iter().position(|n| n == "task_embedding_json");
            while let Some(r) = rows.next().map_err(sql_err)? {
                let res = (|| -> Result<Pathway, PwErr> {
                    let mut row = Row::default();
                    for (i, n) in names.iter().enumerate() {
                        if Some(i) == ei {
                            continue;
                        }
                        row.cols.push((n.clone(), Col::from_ref(r.get_ref(i).map_err(sql_err)?)?));
                    }
                    let id_keep = match row.get("id") {
                        Some(Col::Text(s)) => keep(s),
                        _ => false,
                    };
                    let emb = match ei {
                        Some(i) => Emb::read(r.get_ref(i).map_err(sql_err)?, id_keep)?,
                        None => return Err(PwErr::Unreadable("the row has no task_embedding_json column".into())),
                    };
                    from_row(&row, &emb)
                })();
                out.push(res);
            }
            Ok(out)
        })?;
        let mut all = vec![];
        for part in parts {
            for r in part {
                all.push(r?);
            }
        }
        Ok(all)
    }
}

// ---------------------------------------------------------------------------
// pydantic's JSON
// ---------------------------------------------------------------------------

/// pydantic's text for a float: its own exponent form, and null for NaN
/// and the infinities.
pub fn float_json(f: f64) -> String {
    if !f.is_finite() {
        return "null".into();
    }
    crate::cli::envelope::float_json(f)
}

/// A dict key as pydantic's JSON writes it: a lone surrogate becomes its
/// three UTF-8-style bytes, which read back as three U+FFFD.
fn key_text(k: &str) -> String {
    if !has_surrogate(k) {
        return k.to_string();
    }
    k.chars().map(|c| if as_surrogate(c).is_some() { "\u{FFFD}\u{FFFD}\u{FFFD}".to_string() } else { c.to_string() }).collect()
}

fn pydantic_keys(v: &Value) -> Value {
    match v {
        Value::List(l) => Value::List(l.iter().map(pydantic_keys).collect()),
        Value::Obj(o) => {
            let mut out = Object::new();
            for k in o.keys() {
                out.set(&key_text(k), pydantic_keys(o.value(k)));
            }
            Value::Obj(out)
        }
        other => other.clone(),
    }
}

/// `model_dump_json()`: compact, non-ASCII kept.
pub fn dump_json(v: &Value) -> String {
    let mut b = String::new();
    compact(&mut b, v);
    b
}

fn compact(b: &mut String, v: &Value) {
    match v {
        Value::Float(f) => b.push_str(&float_json(*f)),
        Value::List(l) => {
            b.push('[');
            for (i, e) in l.iter().enumerate() {
                if i > 0 {
                    b.push(',');
                }
                compact(b, e);
            }
            b.push(']');
        }
        Value::Obj(o) => {
            b.push('{');
            for (i, k) in o.keys().iter().enumerate() {
                if i > 0 {
                    b.push(',');
                }
                b.push_str(&dumps(&Value::Str(key_text(k)), false));
                b.push(':');
                compact(b, o.value(k));
            }
            b.push('}');
        }
        other => b.push_str(&dumps(other, false)),
    }
}

/// `model_dump_json(indent=n)`.
pub fn dump_json_indent(v: &Value, n: usize) -> String {
    dumps_indent_with(&pydantic_keys(v), n, false, &float_json)
}

/// `model_dump(mode="json")`: the same values, with NaN and the
/// infinities as None.
pub fn json_mode(v: &Value) -> Value {
    crate::gate::pyjson::any_json(v)
}

/// `json.dumps(v)` with its default ensure_ascii.
pub fn py_dumps(v: &Value) -> String {
    dumps(v, true)
}

/// `json.dumps(v, indent=n)`.
pub fn py_dumps_indent(v: &Value, n: usize) -> String {
    dumps_indent_with(v, n, true, &float_repr)
}
